File: 05-failure/01-utilisation-littles-law-and-the-latency-knee/cpp/latency_knee.cpp
// Layer 5 - Topic 1: the latency knee in C++.
//
// WHAT THIS DEMONSTRATES
//   C++ gives you no runtime, no scheduler and no pool, so the queue is
//   visible as an object you can print. That is this file's contribution:
//   everywhere else in this topic the waiting room is inside somebody's
//   library, and here it is a std::deque you can watch grow.
//
//   The second contribution is the measurement that settles the argument
//   in a real incident. This program asks the kernel, via getrusage(2),
//   how much CPU it actually consumed, and prints it beside wall time. At
//   every row past the knee, latency is multiples of the service time
//   while CPU utilisation is a few percent. If latency climbs while CPU
//   sits low, you are queueing on a COUNT -- pool slots, threads, tokens,
//   workers -- and adding CPU cannot help you.
//
//   (On Linux you would probably reach for /proc/self/stat here. That path
//   does not exist on Darwin. getrusage is POSIX and works on both, which
//   is the general lesson: prefer the syscall to the pseudo-filesystem.)
//
// WHAT TO LOOK FOR IN THE OUTPUT
//  1. `achieved` plateaus at workers / service time: lambda_max = L / W.
//  2. p99 tracks the S/(1-rho) column until the queue stops draining.
//  3. `queue max` is the high-water mark of the waiting room. Nobody
//     bounded it; nothing would have stopped it reaching a million.
//  4. `cpu %` stays low in every row. The machine is not busy. It is full.
//  5. Doubling the worker count moves capacity and the knee proportionally.
//
// BUILD AND RUN
//	c++ -O2 -std=c++17 -pthread -o /tmp/latency_knee latency_knee.cpp && /tmp/latency_knee
#include <sys/resource.h>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <cmath>
#include <condition_variable>
#include <cstdio>
#include <deque>
#include <mutex>
#include <random>
#include <string>
#include <thread>
#include <vector>

using Clock = std::chrono::steady_clock;
using TimePoint = Clock::time_point;

static constexpr auto kService = std::chrono::milliseconds(40);
static constexpr auto kStep = std::chrono::seconds(8);
static constexpr auto kGaugeEvery = std::chrono::milliseconds(20);
static const std::vector<int> kWorkerCounts = {5, 10};
static const std::vector<double> kRhos = {0.2, 0.5, 0.8, 0.9, 0.95, 1.1};

static double ms_since(TimePoint from, TimePoint to) {
    return std::chrono::duration<double, std::milli>(to - from).count();
}

// One request, carrying the time it was SCHEDULED to arrive. Every latency
// in this file is measured from that instant, not from the moment a thread
// got round to enqueueing it. A generator that starts the clock at dispatch
// forgives itself for being late; real users do not. Topic 6.
struct Request {
    TimePoint scheduled;
};

// A bounded number of workers in front of an UNBOUNDED waiting room. This
// is the default shape of almost every thread pool ever written, including
// the one in Layer 1 topic 3, and topic 5 of this layer is about why the
// second half of that sentence is a defect rather than a default.
class Server {
public:
    explicit Server(int workers) {
        for (int i = 0; i < workers; i++) {
            threads_.emplace_back([this] { work(); });
        }
    }

    ~Server() {
        {
            std::lock_guard<std::mutex> lk(m_);
            stop_ = true;
        }
        cv_.notify_all();
        for (auto& t : threads_) t.join();
    }

    void submit(Request r) {
        {
            std::lock_guard<std::mutex> lk(m_);
            queue_.push_back(r);
            queue_max_ = std::max<size_t>(queue_max_, queue_.size());
        }
        inflight_.fetch_add(1);
        cv_.notify_one();
    }

    void reset_samples() {
        std::lock_guard<std::mutex> lk(m_);
        total_.clear();
        wait_.clear();
        completions_.clear();
        queue_max_ = 0;
    }

    struct Snapshot {
        std::vector<double> total;
        std::vector<double> wait;
        std::vector<TimePoint> completions;
        size_t queue_max;
    };

    Snapshot snapshot() {
        std::lock_guard<std::mutex> lk(m_);
        return Snapshot{total_, wait_, completions_, queue_max_};
    }

    int inflight() const { return inflight_.load(); }

    void drain() {
        std::unique_lock<std::mutex> lk(m_);
        idle_.wait(lk, [this] { return queue_.empty() && busy_ == 0; });
    }

private:
    void work() {
        while (true) {
            Request r;
            {
                std::unique_lock<std::mutex> lk(m_);
                cv_.wait(lk, [this] { return stop_ || !queue_.empty(); });
                if (stop_ && queue_.empty()) return;
                r = queue_.front();
                queue_.pop_front();
                busy_++;
            }
            // The instant a worker picks this up is the end of its queue
            // wait and the start of its service. Everything between
            // `scheduled` and here is pure waiting: no code ran for it.
            TimePoint started = Clock::now();
            std::this_thread::sleep_for(kService);
            TimePoint done = Clock::now();
            inflight_.fetch_sub(1);
            {
                std::lock_guard<std::mutex> lk(m_);
                wait_.push_back(std::max(0.0, ms_since(r.scheduled, started)));
                total_.push_back(ms_since(r.scheduled, done));
                completions_.push_back(done);
                busy_--;
                if (queue_.empty() && busy_ == 0) idle_.notify_all();
            }
        }
    }

    std::vector<std::thread> threads_;
    std::deque<Request> queue_;
    std::vector<double> total_, wait_;
    std::vector<TimePoint> completions_;
    size_t queue_max_ = 0;
    int busy_ = 0;
    std::atomic<int> inflight_{0};
    std::mutex m_;
    std::condition_variable cv_, idle_;
    bool stop_ = false;
};

static double percentile(std::vector<double> v, double p) {
    if (v.empty()) return 0.0;
    std::sort(v.begin(), v.end());
    size_t k = static_cast<size_t>(std::lround(p / 100.0 * (v.size() - 1)));
    return v[std::min(k, v.size() - 1)];
}

static double mean(const std::vector<double>& v) {
    if (v.empty()) return 0.0;
    double s = 0;
    for (double x : v) s += x;
    return s / v.size();
}

// Total CPU seconds this process has burned, user + system, straight from
// the kernel. Subtract two readings to get the CPU consumed by a step.
static double cpu_seconds() {
    struct rusage ru;
    getrusage(RUSAGE_SELF, &ru);
    return ru.ru_utime.tv_sec + ru.ru_utime.tv_usec / 1e6 +
           ru.ru_stime.tv_sec + ru.ru_stime.tv_usec / 1e6;
}

struct StepResult {
    double offered, achieved, p50, p99, wait_p50, mean_total, gauge_l, cpu_percent;
    size_t queue_max;
};

// Exponential gaps make a Poisson process: the standard model for
// independent users arriving. Evenly spaced arrivals would understate the
// queue, because bursts are what fill it.
static StepResult step(Server& server, double rate, std::chrono::seconds dur, unsigned seed) {
    std::mt19937 rng(seed);
    std::exponential_distribution<double> gap(rate);

    server.reset_samples();
    std::vector<double> gauge;
    std::atomic<bool> sampling{true};
    std::thread sampler([&] {
        while (sampling.load()) {
            std::this_thread::sleep_for(kGaugeEvery);
            gauge.push_back(server.inflight());
        }
    });

    TimePoint begin = Clock::now();
    TimePoint deadline = begin + dur;
    double cpu_before = cpu_seconds();

    size_t sent = 0;
    TimePoint at = begin;
    while (true) {
        at += std::chrono::duration_cast<Clock::duration>(
            std::chrono::duration<double>(gap(rng)));
        if (at > deadline) break;
        std::this_thread::sleep_until(at);
        sent++;
        server.submit(Request{at});
    }

    // Drain. Past rho=1 this is where the backlog built up during the step
    // finally comes out, which is why those rows carry latencies larger
    // than the step itself.
    server.drain();
    double cpu_used = cpu_seconds() - cpu_before;
    sampling.store(false);
    sampler.join();

    auto snap = server.snapshot();
    double seconds = std::chrono::duration<double>(dur).count();
    size_t in_window = 0;
    for (auto& c : snap.completions) {
        if (c <= deadline) in_window++;
    }
    double wall = std::chrono::duration<double>(Clock::now() - begin).count();
    unsigned cores = std::max(1u, std::thread::hardware_concurrency());

    return StepResult{
        static_cast<double>(sent) / seconds,
        static_cast<double>(in_window) / seconds,
        percentile(snap.total, 50),
        percentile(snap.total, 99),
        percentile(snap.wait, 50),
        // Little's Law is a statement about MEANS. L = lambda * p50 is not
        // a law and stops holding exactly when the distribution skews,
        // which is exactly when you reach for it.
        mean(snap.total),
        mean(gauge),
        100.0 * cpu_used / (wall * cores),
        snap.queue_max,
    };
}

static const char* kHeader =
    "  rho   offered  achieved      p50      p99   wait p50   L (gauge)   lam*Wbar   S/(1-rho)   queue max   cpu %";

static void print_row(double rho, const StepResult& r, double service) {
    char predicted[16];
    if (rho < 1.0) {
        std::snprintf(predicted, sizeof(predicted), "%9.1f", service / (1.0 - rho) * 1000.0);
    } else {
        std::snprintf(predicted, sizeof(predicted), "%9s", "inf");
    }
    std::printf("%5.2f %9.1f %9.1f %8.1f %8.1f %10.1f %11.1f %10.1f %s %11zu %7.1f\n",
                rho, r.offered, r.achieved, r.p50, r.p99, r.wait_p50, r.gauge_l,
                r.achieved * r.mean_total / 1000.0, predicted, r.queue_max, r.cpu_percent);
}

static void chart(const std::vector<double>& p99s) {
    double top = 1.0;
    for (double v : p99s) top = std::max(top, v);
    std::printf("\n  p99 (ms) against rho\n");
    for (size_t i = 0; i < p99s.size(); i++) {
        int n = std::max(1, static_cast<int>(std::lround(56 * p99s[i] / top)));
        std::printf("  rho=%-6.2f|%s %.0f\n", kRhos[i], std::string(n, '#').c_str(), p99s[i]);
    }
    std::printf("  %10s+%s %.0f ms full scale\n", "", std::string(56, '-').c_str(), top);
}

static void sweep(int workers) {
    Server server(workers);

    // Measure S rather than assuming kService. A capacity computed from a
    // constant nobody measured is the commonest way this experiment lies.
    StepResult warm = step(server, 5.0, std::chrono::seconds(2), 12345);
    double service = warm.mean_total / 1000.0;
    double capacity = workers / service;

    std::printf("\n=== %d workers, measured service time S = %.1f ms ===\n", workers, service * 1000);
    std::printf("predicted capacity L/S = %.1f rps\n\n", capacity);
    std::printf("%s\n", kHeader);
    std::printf("%s\n", std::string(std::char_traits<char>::length(kHeader), '-').c_str());

    std::vector<double> p99s;
    for (size_t i = 0; i < kRhos.size(); i++) {
        StepResult r = step(server, capacity * kRhos[i], kStep, 777 + i * 31);
        print_row(kRhos[i], r, service);
        p99s.push_back(r.p99);
    }
    chart(p99s);
}

int main() {
    std::printf("Latency knee in C++: a fixed worker pool, an unbounded waiting room,\n");
    std::printf("and open-model Poisson arrivals. %u hardware threads available.\n",
                std::max(1u, std::thread::hardware_concurrency()));
    for (int w : kWorkerCounts) sweep(w);
    std::printf("\nThe two sweeps ran identical code and an identical ramp; the only\n");
    std::printf("difference is the worker count. Compare their capacity lines and their\n");
    std::printf("rho=0.9 rows: that is the whole topic in two numbers.\n");
    std::printf("\nThen read the cpu %% column top to bottom. Not one row is CPU-bound.\n");
    std::printf("Every millisecond of that latency was spent waiting for a count.\n");
    return 0;
}

File: 05-failure/01-utilisation-littles-law-and-the-latency-knee/java/LatencyKnee.java
// Layer 5 - Topic 1: the latency knee in Java, and the counterintuitive
// result about virtual threads.
//
// WHAT THIS DEMONSTRATES
//
//   Java stacks two counts and most teams tune the wrong one. The servlet
//   container's thread pool (Tomcat's maxThreads, 200 by default under
//   Spring Boot) is the number people raise; HikariCP's maximumPoolSize
//   (default 10) is almost always the smaller number and therefore the
//   actual ceiling. Raising maxThreads moves nothing, because lambda_max
//   is L/S where L is the SMALLEST count on the path.
//
//   The second, less intuitive half: switching the container to virtual
//   threads makes the knee SHARPER, not gentler. A bounded platform pool
//   with a bounded queue eventually *rejects*, and a rejection is fast --
//   it caps latency by discarding work. Virtual threads remove that
//   ceiling, so nothing is ever rejected and every excess request queues
//   on the connection pool instead. You traded a bounded-latency failure
//   for an unbounded-latency one. Predict which you want before you look.
//
// WHAT TO LOOK FOR IN THE OUTPUT
//   1. `achieved` plateaus at pool / S, not at containerThreads / S. The
//      200-thread container is irrelevant to capacity.
//   2. p99 tracks the S/(1-rho) column while rho < 1 and leaves it behind
//      once the queue stops draining.
//   3. `wait p50` -- time queued for a connection, not time spent using
//      one -- is ~0 at rho=0.2 and is nearly all of the latency by 0.95.
//      The handler never got slower.
//   4. Doubling maximumPoolSize moves capacity and the knee with it.
//   5. Platform sweep at rho=1.1 has a non-zero `rejected` column and a
//      finite p99. Virtual sweep at rho=1.1 has zero rejections and a p99
//      several times larger. Same pool, same load, same service time.
//
// The load generator is OPEN MODEL: arrivals are a Poisson process
// computed ahead of time, and each request's clock starts when it was
// SCHEDULED to arrive, not when this program managed to dispatch it. A
// generator that starts the clock at dispatch forgives itself for being
// late. Topic 6 is about why that matters.
//
// RUN
//   javac LatencyKnee.java -d /tmp/javabuild && java -cp /tmp/javabuild LatencyKnee
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.Future;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.Semaphore;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

public class LatencyKnee {

    // ------------------------------------------------------------ config

    static final long SERVICE_NANOS = 40_000_000L;   // 40ms inside a connection
    static final double STEP_SECONDS = 8.0;
    static final double[] RHOS = {0.2, 0.5, 0.8, 0.9, 0.95, 1.1};
    static final int[] POOL_SIZES = {5, 10};         // HikariCP maximumPoolSize
    static final int CONTAINER_THREADS = 200;        // Tomcat maxThreads default
    static final int CONTAINER_QUEUE = 100;          // Tomcat acceptCount default
    static final double OVERLOAD_RHO = 1.5;          // sustained, not a spike
    static final long GAUGE_EVERY_NANOS = 20_000_000L;

    static final Random RNG = new Random(20250501L);

    // ------------------------------------------------- the bounded thing

    /**
     * Stands in for HikariCP plus the database behind it. Acquiring a permit
     * is `getConnection()`; holding it for SERVICE_NANOS is the query. This
     * is the only genuinely bounded resource in the program, which is the
     * entire point of the topic: capacity is a property of the smallest
     * count, and everything else is decoration.
     */
    static final class ConnectionPool {
        private final Semaphore permits;

        ConnectionPool(int size) {
            // Fair, because HikariCP hands connections out in arrival order
            // and an unfair pool would give you a tail that is an artefact
            // of the lock rather than of the queue.
            this.permits = new Semaphore(size, true);
        }

        void query() throws InterruptedException {
            permits.acquire();
            try {
                sleepNanos(SERVICE_NANOS);
            } finally {
                permits.release();
            }
        }
    }

    // --------------------------------------------------- the two carriers

    interface Joinable {
        void join() throws InterruptedException;
    }

    /** How a request gets a thread. The only difference between the sweeps. */
    interface Carrier {
        /** Returns a handle, or null if the request was rejected outright. */
        Joinable submit(Runnable r);

        void shutdown();

        String describe();
    }

    /**
     * The traditional servlet container: a fixed pool of platform threads
     * with a bounded queue in front of it. When both are full the request
     * is REJECTED -- which is a real answer, delivered fast, and is why
     * this configuration's p99 stays finite past rho=1.
     */
    static Carrier platformCarrier() {
        ThreadPoolExecutor ex = new ThreadPoolExecutor(
                CONTAINER_THREADS, CONTAINER_THREADS,
                0L, TimeUnit.MILLISECONDS,
                new ArrayBlockingQueue<>(CONTAINER_QUEUE),
                new ThreadPoolExecutor.AbortPolicy());
        ex.prestartAllCoreThreads();
        return new Carrier() {
            public Joinable submit(Runnable r) {
                try {
                    Future<?> f = ex.submit(r);
                    return () -> {
                        try {
                            f.get();
                        } catch (Exception ignored) {
                        }
                    };
                } catch (RejectedExecutionException e) {
                    return null;   // 503 at the container, before any Java of yours runs
                }
            }

            public void shutdown() {
                ex.shutdownNow();
            }

            public String describe() {
                return CONTAINER_THREADS + " platform threads + queue of " + CONTAINER_QUEUE
                        + " (rejects when both are full)";
            }
        };
    }

    /**
     * Java 21 virtual threads: one per request, no ceiling worth the name.
     * Nothing is ever rejected, so nothing caps latency except the pool.
     */
    static Carrier virtualCarrier() {
        return new Carrier() {
            public Joinable submit(Runnable r) {
                Thread t = Thread.ofVirtual().start(r);
                return t::join;
            }

            public void shutdown() {
            }

            public String describe() {
                return "one virtual thread per request (no ceiling, nothing is ever rejected)";
            }
        };
    }

    // ------------------------------------------------------- measurement

    record Result(double offered, double achieved, double p50, double p99,
                  double waitP50, double meanTotal, double gaugeL, long rejected) {
    }

    /**
     * One measurement step at a fixed offered rate.
     */
    static Result step(ConnectionPool pool, Carrier carrier, double rate, double durSeconds)
            throws InterruptedException {
        List<Double> total = Collections.synchronizedList(new ArrayList<>());
        List<Long> completions = Collections.synchronizedList(new ArrayList<>());
        List<Double> gauge = Collections.synchronizedList(new ArrayList<>());
        AtomicInteger inflight = new AtomicInteger();
        AtomicLong rejected = new AtomicLong();
        List<Joinable> handles = new ArrayList<>();

        long begin = System.nanoTime();
        long deadline = begin + (long) (durSeconds * 1e9);

        Thread sampler = Thread.ofPlatform().daemon().start(() -> {
            try {
                while (!Thread.currentThread().isInterrupted()) {
                    sleepNanos(GAUGE_EVERY_NANOS);
                    gauge.add((double) inflight.get());
                }
            } catch (InterruptedException ignored) {
            }
        });

        int sent = 0;
        long at = begin;
        while (true) {
            at += (long) (expovariate() / rate * 1e9);
            if (at > deadline) break;
            long wait = at - System.nanoTime();
            if (wait > 0) sleepNanos(wait);
            sent++;
            final long scheduled = at;

            Joinable h = carrier.submit(() -> {
                inflight.incrementAndGet();
                try {
                    pool.query();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                } finally {
                    inflight.decrementAndGet();
                }
                long done = System.nanoTime();
                // Latency measured from the SCHEDULED arrival, so lateness in
                // the generator counts against the system under test rather
                // than being quietly forgiven.
                total.add((done - scheduled) / 1e6);
                completions.add(done);
            });

            if (h == null) {
                rejected.incrementAndGet();
            } else {
                handles.add(h);
            }
        }

        // Drain. Past rho=1 this is where the backlog built up during the
        // step finally comes out, which is why those rows carry latencies
        // longer than the step itself.
        for (Joinable h : handles) h.join();
        sampler.interrupt();

        // Queue wait = total latency minus the service time we know each
        // request spent actually holding a connection. With a single
        // controlled S this subtraction is exact, and it isolates the term
        // that the 1/(1-rho) curve is actually about: the handler never got
        // slower, only the waiting in front of it did.
        List<Double> sortedTotal = new ArrayList<>(total);
        List<Double> sortedWaits = new ArrayList<>(sortedTotal.size());
        for (double t : sortedTotal) {
            sortedWaits.add(Math.max(0.0, t - SERVICE_NANOS / 1e6));
        }
        Collections.sort(sortedTotal);
        Collections.sort(sortedWaits);

        long inWindow = completions.stream().filter(c -> c <= deadline).count();
        return new Result(
                sent / durSeconds,
                inWindow / durSeconds,
                percentile(sortedTotal, 50),
                percentile(sortedTotal, 99),
                percentile(sortedWaits, 50),
                mean(sortedTotal),
                mean(gauge),
                rejected.get());
    }

    // ------------------------------------------------------------ sweeps

    static final String HEADER =
            "  rho   offered  achieved      p50      p99   wait p50   L (gauge)   lam*Wbar   S/(1-rho)  rejected";

    static void printRow(double rho, Result r, double serviceSeconds) {
        String predicted = rho < 1
                ? String.format("%9.1f", serviceSeconds / (1 - rho) * 1000)
                : "      inf";
        System.out.printf("%5.2f %9.1f %9.1f %8.1f %8.1f %10.1f %11.1f %10.1f %s %9d%n",
                rho, r.offered(), r.achieved(), r.p50(), r.p99(), r.waitP50(), r.gaugeL(),
                r.achieved() * r.meanTotal() / 1000, predicted, r.rejected());
    }

    static double[] sweep(int poolSize, Carrier carrier, String label) throws InterruptedException {
        ConnectionPool pool = new ConnectionPool(poolSize);

        // Measure S rather than assuming it. A capacity computed from a
        // constant nobody measured is the commonest way this experiment
        // quietly lies to you.
        Result warm = step(pool, carrier, 5, 2.0);
        double service = warm.meanTotal() / 1000;
        double capacity = poolSize / service;

        System.out.printf("%n=== %s ===%n", label);
        System.out.printf("carrier: %s%n", carrier.describe());
        System.out.printf("maximumPoolSize = %d, measured service time S = %.1f ms%n",
                poolSize, service * 1000);
        System.out.printf("predicted capacity L/S = %.1f rps%n%n", capacity);
        System.out.println(HEADER);
        System.out.println("-".repeat(HEADER.length()));

        double[] p99s = new double[RHOS.length];
        for (int i = 0; i < RHOS.length; i++) {
            Result r = step(pool, carrier, capacity * RHOS[i], STEP_SECONDS);
            printRow(RHOS[i], r, service);
            p99s[i] = r.p99();
        }
        return p99s;
    }

    /** The knee is a shape, and a table of numbers hides shapes. */
    static void chart(double[] p99s) {
        double top = Arrays.stream(p99s).max().orElse(1);
        if (top == 0) top = 1;
        System.out.println("\n  p99 (ms) against rho");
        for (int i = 0; i < p99s.length; i++) {
            int n = Math.max(1, (int) Math.round(56 * p99s[i] / top));
            System.out.printf("  rho=%-6.2f|%s %.0f%n", RHOS[i], "#".repeat(n), p99s[i]);
        }
        System.out.printf("  %10s+%s %.0f ms full scale%n", "", "-".repeat(56), top);
    }

    // -------------------------------------------------------------- main

    public static void main(String[] args) throws Exception {
        System.out.printf("Latency knee in Java (%s, %d cores).%n",
                Runtime.version(), Runtime.getRuntime().availableProcessors());
        System.out.println("Two counts stacked: a " + CONTAINER_THREADS + "-thread container in front of a");
        System.out.println("connection pool of " + POOL_SIZES[0] + " then " + POOL_SIZES[1]
                + ". Watch which one sets capacity.");

        double[] last = null;
        for (int size : POOL_SIZES) {
            Carrier platform = platformCarrier();
            last = sweep(size, platform, "Platform threads, maximumPoolSize = " + size);
            platform.shutdown();
            chart(last);
        }

        Carrier virtual = virtualCarrier();
        int size = POOL_SIZES[POOL_SIZES.length - 1];
        double[] virtualP99 = sweep(size, virtual, "Virtual threads, maximumPoolSize = " + size);
        virtual.shutdown();
        chart(virtualP99);

        overloadComparison(size);

        System.out.println();
        System.out.println("Raising maxThreads did nothing to capacity in any sweep, because");
        System.out.println("capacity is L/S at the SMALLEST count and that was never the threads.");
        System.out.println("Removing the thread bound entirely did change something, and not in");
        System.out.println("the direction the phrase 'removed a bottleneck' suggests: the bounded");
        System.out.println("container was shedding load for you, and you did not know it was a");
        System.out.println("feature until you deleted it. Topic 5 is about doing that on purpose.");
    }

    /**
     * The sweeps above stop at rho=1.1, where 8 seconds of 10% excess is not
     * enough backlog to reach any container bound. Sustained overload is
     * where the two carriers stop agreeing, so run one on purpose: the same
     * pool, the same offered rate, only the thread carrier differs.
     */
    static void overloadComparison(int poolSize) throws InterruptedException {
        System.out.printf("%n=== Sustained overload: rho = %.1f, maximumPoolSize = %d ===%n",
                OVERLOAD_RHO, poolSize);
        System.out.println("Same pool, same offered rate, same handler. Only the carrier differs.\n");
        System.out.printf("%-22s %9s %9s %9s %9s %10s%n",
                "carrier", "offered", "achieved", "p50", "p99", "rejected");
        System.out.println("-".repeat(73));

        double capacity = poolSize / (SERVICE_NANOS / 1e9);
        for (String kind : new String[]{"platform", "virtual"}) {
            Carrier c = kind.equals("platform") ? platformCarrier() : virtualCarrier();
            ConnectionPool pool = new ConnectionPool(poolSize);
            Result r = step(pool, c, capacity * OVERLOAD_RHO, STEP_SECONDS);
            c.shutdown();
            System.out.printf("%-22s %9.1f %9.1f %9.1f %9.1f %10d%n",
                    kind, r.offered(), r.achieved(), r.p50(), r.p99(), r.rejected());
        }

        System.out.println();
        System.out.println("  The platform row rejects and keeps a finite p99. The virtual row");
        System.out.println("  rejects nothing and pays for it in the tail, because the backlog");
        System.out.println("  has nowhere to go but the queue in front of the pool. Neither is");
        System.out.println("  'correct' -- but only one of them was a decision somebody made.");
    }

    // ---------------------------------------------------------- plumbing

    static double expovariate() {
        return -Math.log(1 - RNG.nextDouble());
    }

    static void sleepNanos(long nanos) throws InterruptedException {
        if (nanos <= 0) return;
        Thread.sleep(nanos / 1_000_000L, (int) (nanos % 1_000_000L));
    }

    static double percentile(List<Double> sorted, double p) {
        if (sorted.isEmpty()) return 0;
        int k = (int) Math.round(p / 100 * (sorted.size() - 1));
        return sorted.get(Math.min(k, sorted.size() - 1));
    }

    static double mean(List<Double> v) {
        if (v.isEmpty()) return 0;
        double s = 0;
        for (double x : v) s += x;
        return s / v.size();
    }
}
